solo-chains/runtime/dancelight/src/lib.rs

```diff
@@ -2909,54 +2911,127 @@ impl frame_support::traits::Randomness<Hash, BlockNumber> for BabeCurrentBlockRa
+                        assignment_imbalance,
+                    );
+                }
+                return Err(e);
             }
+        }
 
-            let remaining_block_credits = block_credits_needed.saturating_sub(free_block_credits);
-            let remaining_session_credits = 1u32.saturating_sub(free_session_credits);
+        if let Some(assignment_imbalance) = maybe_assignment_imbalance {
+            <Runtime as pallet_services_payment::Config>::OnChargeForCollatorAssignment::on_unbalanced(assignment_imbalance);
+        }
 
-            let (block_production_costs, _) = <Runtime as pallet_services_payment::Config>::ProvideBlockProductionCost::block_cost(para_id);
-            let (collator_assignment_costs, _) = <Runtime as pallet_services_payment::Config>::ProvideCollatorAssignmentCost::collator_assignment_cost(para_id);
-            // let's check if we can withdraw
-            let remaining_block_credits_to_pay = u128::from(remaining_block_credits).saturating_mul(block_production_costs);
-            let remaining_session_credits_to_pay = u128::from(remaining_session_credits).saturating_mul(collator_assignment_costs);
+        // If the para has been assigned collators for this session it must have enough block credits
+        // for the current and the next session.
+        let block_credits_needed = if currently_assigned.contains(&para_id) {
+            blocks_per_session * 2
+        } else {
+            blocks_per_session
+        };
+        // Check if the container chain has enough credits for producing blocks
+        let free_block_credits =
+            pallet_services_payment::BlockProductionCredits::<Runtime>::get(para_id)
+                .unwrap_or_default();
+        let remaining_block_credits = block_credits_needed.saturating_sub(free_block_credits);
+        let (block_production_costs, _) =
+            <Runtime as pallet_services_payment::Config>::ProvideBlockProductionCost::block_cost(
+                &para_id,
+            );
+        // Check if we can withdraw
+        let remaining_block_credits_to_pay =
+            u128::from(remaining_block_credits).saturating_mul(block_production_costs);
+        let remaining_to_pay = remaining_block_credits_to_pay;
+        // This should take into account whether we tank goes below ED
+        // The true refers to keepAlive
+        Balances::can_withdraw(
+            &pallet_services_payment::Pallet::<Runtime>::parachain_tank(para_id),
+            remaining_to_pay,
+        )
+        .into_result(true)?;
+        // TODO: Have proper weight
+        Ok(Weight::zero())
+    }
+}
 
-            let remaining_to_pay = remaining_block_credits_to_pay.saturating_add(remaining_session_credits_to_pay).saturating_add(max_tip);
+impl<AC> RemoveParaIdsWithNoCredits<BalanceOf<Runtime>, AC> for RemoveParaIdsWithNoCreditsImpl {
+    fn pre_assignment_remove_para_ids_with_no_credits(
```

Comment:
Here I would rename `pre_assignment_remove_para_ids_with_no_credits` to simply `pre_assignment_hook` or  `pre_assignment_filtering`